llama-index-core/llama_index/core/tools/function_tool.py

```diff
@@ -108,20 +144,30 @@ def async_fn(self) -> AsyncCallable:
         return self._async_fn
 
     def call(self, *args: Any, **kwargs: Any) -> ToolOutput:
-        """Call."""
+        """Sync Call."""
         tool_output = self._fn(*args, **kwargs)
+        final_output_content = str(tool_output)
+        # Execute sync callback, if available
+        callback_output = self._run_sync_callback(tool_output)
+        if callback_output:
+            final_output_content += f" Callback: {callback_output}"
```

Comment:
Actually, looking at this again, this seems rather opinionated 😅 What's the intended use case for this? 

I kind of expected the callback to just either be logging something, or outright modifying the tool output in place.

Appending a string like this seems kind of strange

Reply:
I'll explain the use case we applied here to see if it becomes a bit clearer:

I have a tool that sends information to the external system of my company.
The agent is responsible for setting all parameters for the API call, but I can't blindly trust the information constructed by the agent.
So, I use this callback strategy to request user confirmation.
This confirmation returns to the agent's interaction, allowing the agent to decide the next steps.
Here, we apply this rule with interactions between agents and the front end for confirmation, but if you check the example notebook I attached, you can see this flow abstracted in a certain way that's easier to understand.

Basically, I need the callback return to influence the remainder of the agent's flow. I couldn't think of an easier way to adapt the classes without violating any principles. If the user doesn't want the callback to influence the flow, they just don't return anything in the function.

Reply:
It seems like to me a better design might be 

```
callback_output: ToolOutput | None = self._run_sync_callback(tool_output)
```

Basically the callback either returns a new tool output to override the existing, or returns None and the original tool_output is used

Reply:
Then its up to the user how the callback changes the tool output